core/src/main/scala/org/apache/spark/executor/ExecutorMetricsPoller.scala

```diff
@@ -0,0 +1,202 @@
+
+    // Put a new entry in stageTCMP for the stage if there isn't one already.
+    // Increment the task count.
+    val (count, _) = stageTCMP.computeIfAbsent((stageId, stageAttemptId),
+      _ => (new AtomicLong(0), new AtomicLongArray(ExecutorMetricType.numMetrics)))
+    val stageCount = count.incrementAndGet()
+    logDebug(s"stageTCMP: ($stageId, $stageAttemptId) -> $stageCount")
+  }
+
+  /**
+   * Called by TaskRunner#run. It should only be called if onTaskStart has been called with
+   * the same arguments.
+   *
+   * @param taskId the id of the task that was run.
+   * @param stageId the id of the stage the task belongs to.
+   * @param stageAttemptId the attempt number of the stage the task belongs to.
+   */
+  def onTaskCompletion(taskId: Long, stageId: Int, stageAttemptId: Int): Unit = {
+    // Decrement the task count.
+    // Remove the entry from stageTCMP if the task count reaches zero.
+
+    def decrementCount(stage: StageKey, countAndPeaks: TCMP): TCMP = {
+      val count = countAndPeaks._1
+      val countValue = count.decrementAndGet()
+      if (countValue == 0L) {
+        logDebug(s"removing (${stage._1}, ${stage._2}) from stageTCMP")
+        null
+      } else {
+        logDebug(s"stageTCMP: (${stage._1}, ${stage._2}) -> " + countValue)
+        countAndPeaks
+      }
+    }
+
+    stageTCMP.computeIfPresent((stageId, stageAttemptId), decrementCount)
+
+    // Remove the entry from taskMetricPeaks for the task.
+    taskMetricPeaks.remove(taskId)
+  }
+
+  /**
+   * Called by TaskRunner#run.
+   *
+   * @param taskId the id of the task that was run.
+   */
+  def getTaskMetricPeaks(taskId: Long): Array[Long] = {
+    // If this is called with an invalid taskId or a valid taskId but the task was killed and
+    // onTaskStart was therefore not called, then we return an array of zeros.
+    val currentPeaks = taskMetricPeaks.get(taskId) // may be null
+    val metricPeaks = new Array[Long](ExecutorMetricType.numMetrics) // initialized to zeros
+    if (currentPeaks != null) {
+      ExecutorMetricType.metricToOffset.foreach { case (_, i) =>
+        metricPeaks(i) = currentPeaks.get(i)
+      }
+    }
+    metricPeaks
+  }
+
+
+  /**
+   * Called by the reportHeartBeat function defined in Executor and passed to its Heartbeater.
+   * It resets the metric peaks in stageTCMP before returning the executor updates.
+   * Thus, the executor updates contains the per-stage metric peaks since the last heartbeat
+   * (the last time this method was called).
+   */
+  def getExecutorUpdates(): HashMap[StageKey, ExecutorMetrics] = {
+    // build the executor level memory metrics
+    val executorUpdates = new HashMap[StageKey, ExecutorMetrics]
+    stageTCMP.forEach((k, v) => executorUpdates.put(k, new ExecutorMetrics(v._2)))
+
+    // reset the peaks
+    def resetPeaks(k: StageKey, v: TCMP): TCMP =
+      (v._1, new AtomicLongArray(ExecutorMetricType.numMetrics))
```

Comment:
I think you should reset the peaks within the above `forEach`.  First, it avoids a small race, as otherwise you might grab the last saved value of the metrics, then another thread updates them to new peaks, and then you clear them in this thread.  (its not horrible to lose that update, but if we can avoid it we should.)  Second its best to avoid too many passes over these structures.

Reply:
Combined extracting the metrics and resetting them in one pass.